common/src/primitives/merkle/tree/tree_size.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2021-2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    fmt::{Display, Formatter},
+    num::NonZeroUsize,
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TreeSize(usize);
+
+const MAX_TREE_SIZE: usize = 1 << 31;
+
+#[derive(thiserror::Error, Debug, Clone, PartialEq, Eq)]
+pub enum TreeSizeError {
+    #[error("Zero is invalid size for tree")]
+    ZeroSize,
+    #[error("Tree size must be power of two minus one; this value was found: {0}")]
+    InvalidSize(usize),
+    #[error("Tree with this huge size is not supported: {0}")]
```

Comment:
I feel this is could be a little more useful.

Reply:
I'm not sure I understand. Could you elaborate?